File: examples/flask/thin_middleware.py
# Flask thin-wrapper example (ASH v1.0.0)
# Must use raw request.data bytes.

import time

from flask import Flask, request, jsonify
from ash.core.proof import ash_verify_proof, ash_build_proof_hmac, ash_derive_client_secret
from ash.core.canonicalize import ash_normalize_binding, ash_hash_body, ash_canonicalize_json

app = Flask(__name__)


class HeaderMapView:
    def __init__(self, headers):
        self._headers = headers

    def get_all_ci(self, name: str):
        # Flask headers are case-insensitive; may have multiple values via getlist
        try:
            return self._headers.getlist(name)
        except Exception:
            v = self._headers.get(name)
            return [] if v is None else [v]


def lookup_context(context_id):
    """Look up a stored ASH context by ID.

    In production, this should retrieve the context from a persistent store
    (e.g., Redis) and return a dict with 'nonce' and 'binding' keys,
    or None if the context is not found or expired.
    """
    # TODO: Replace with actual context store lookup
    return None


@app.before_request
def ash_verify():
    # 1. Extract ASH headers
    context_id = request.headers.get("X-ASH-Context-ID")
    client_proof = request.headers.get("X-ASH-Proof")
    timestamp = request.headers.get("X-ASH-Timestamp")

    if not context_id or not client_proof or not timestamp:
        resp = jsonify({"code": "ASH_PROOF_INVALID", "message": "Missing required ASH headers"})
        resp.status_code = 460
        return resp

    # 2. Look up context
    stored = lookup_context(context_id)
    if not stored or not stored.get("nonce"):
        resp = jsonify({"code": "ASH_PROOF_INVALID", "message": "Unknown or expired context"})
        resp.status_code = 460
        return resp

    nonce = stored["nonce"]

    # 3. Normalize binding from the incoming request
    binding = ash_normalize_binding(
        request.method,
        request.path,
        request.query_string.decode("utf-8", errors="replace"),
    )

    # 4. Hash the canonical body
    body_hash = ash_hash_body(
        ash_canonicalize_json(request.get_data(as_text=True) or "")
    )

    # 5. Verify the proof
    valid = ash_verify_proof(nonce, context_id, binding, timestamp, body_hash, client_proof)

    if valid:
        # NOTE: In production, replay protection (context consumption) should be
        # implemented here to prevent the same context from being reused.
        return None

    resp = jsonify({"code": "ASH_PROOF_INVALID", "message": "Proof verification failed"})
    resp.status_code = 460
    return resp

File: examples/express/thin_middleware.js
// Express thin-wrapper middleware (ASH v1.0.0)
// Verifies ASH HMAC proofs on incoming requests using @3maem/ash-node-sdk.

import {
  ashVerifyProof,
  ashNormalizeBinding,
  ashHashBody,
  ashCanonicalizeJson,
} from "@3maem/ash-node-sdk";
import { Buffer } from "buffer";

/**
 * ASH verification middleware for Express.
 *
 * Requires:
 * - A `lookupContext(contextId)` function that returns { nonce, binding }
 *   for a stored ASH context, or null if not found.
 * - `express.raw({ type: '*/*' })` or equivalent raw-body parser applied
 *   before this middleware so that `req.body` is a Buffer.
 *
 * @param {object} opts
 * @param {function} opts.lookupContext - async (contextId) => { nonce, binding } | null
 * @param {number}  [opts.maxSkewSecs=300] - unused here; timestamp skew should be
 *   enforced by the context store or by ashVerifyProof if supported.
 */
export default function ashMiddleware({ lookupContext, maxSkewSecs = 300 } = {}) {
  if (typeof lookupContext !== "function") {
    throw new Error("ashMiddleware requires a lookupContext(contextId) function");
  }

  return async function (req, res, next) {
    // 1. Extract ASH headers
    const contextId = req.headers["x-ash-context-id"];
    const proof     = req.headers["x-ash-proof"];
    const timestamp = req.headers["x-ash-timestamp"];

    if (!contextId || !proof || !timestamp) {
      return res.status(460).json({
        code: "ASH_PROOF_INVALID",
        message: "Missing required ASH headers (X-ASH-Context-ID, X-ASH-Proof, X-ASH-Timestamp)",
      });
    }

    // 2. Look up the stored context (nonce + binding)
    let stored;
    try {
      stored = await lookupContext(contextId);
    } catch (err) {
      return res.status(460).json({
        code: "ASH_PROOF_INVALID",
        message: "Context lookup failed",
      });
    }

    if (!stored || !stored.nonce) {
      return res.status(460).json({
        code: "ASH_PROOF_INVALID",
        message: "Unknown or expired context",
      });
    }

    // 3. Compute binding and body hash from the incoming request
    const rawQuery = req.originalUrl.split("?")[1] || "";
    const binding  = ashNormalizeBinding(req.method, req.path, rawQuery);

    const body      = req.body ? Buffer.from(req.body) : Buffer.alloc(0);
    const bodyStr   = body.length > 0 ? body.toString("utf-8") : "";
    const canonical = ashCanonicalizeJson(JSON.stringify(req.body));
    const bodyHash  = ashHashBody(canonical);

    // 4. Verify the proof
    //    ashVerifyProof(nonce, contextId, binding, timestamp, bodyHash, clientProof)
    try {
      const valid = ashVerifyProof(
        stored.nonce,
        contextId,
        binding,
        timestamp,
        bodyHash,
        proof,
      );

      if (valid) {
        return next();
      }
    } catch (e) {
      // ashVerifyProof throws structured { code, http_status, message } on error
      const status = (e && e.http_status) || 460;
      return res.status(status).json({
        code: (e && e.code) || "ASH_PROOF_INVALID",
        message: (e && e.message) || "Proof verification failed",
      });
    }

    // Proof did not match
    return res.status(460).json({
      code: "ASH_PROOF_INVALID",
      message: "Proof verification failed",
    });

    // NOTE: In production, replay protection should be implemented here
    // by consuming the context after successful verification (e.g., marking
    // it as used in the context store to prevent replay attacks).
  };
}
